### PR TITLE
bugfix:fix the failure to update cluster list dynamically when use eureka of the develop branch

### LINKED ISSUE #4488: develop分支，Eureka做注册中心，修改事务分组的集群名，并不会去获取最新的集群IP列表
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
develop分支，Eureka做注册中心，修改事务分组的集群名，并不会去获取最新的集群IP列表
EurekaRegistryServiceImpl 132行的refreshCluster方法，触发CacheRefreshedEvent事件时执行，参数还是初始化的clusterUpperName,而不会是128行获取的clusterUpperName

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

### PR DESCRIPTION
**Pull request by tuwenlin**
fix the failure to update cluster list dynamically when use eureka of the develop branch
修复develop分支下eureka注册中心无法动态更新服务列表的问题

- [x] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅱ. Does this pull request fix one issue?

fixes #4488

### CODE REVIEW

**Review by slievrly — COMMENTED** [review]
Does that solve the problem? I think you should solve the problem from `subscribeListener`.

**Comment by tuwenlin** [comment]
> subscribeListener

this is a good suggestion，I will try it ,thank you！

File: discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java [hunk 1]
```diff
@@ -126,17 +134,21 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
             return null;
         }
         String clusterUpperName = clusterName.toUpperCase();
-        if (!subscribeListener) {
-            refreshCluster(clusterUpperName);
-            subscribe(null, event -> {
-                try {
+        if (!LISTENER_SERVICE_MAP.containsKey(clusterUpperName)) {
+            Object lock = CLUSTER_LOCK.putIfAbsent(clusterName, new Object());
```
Comment (a364176773) [inline discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java:138]: use computeIfAbsent

Comment (tuwenlin) [reply discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java]: I already did it

File: discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java [hunk 2]
```diff
@@ -126,17 +134,21 @@ public List<InetSocketAddress> lookup(String key) throws Exception {
             return null;
         }
         String clusterUpperName = clusterName.toUpperCase();
-        if (!subscribeListener) {
-            refreshCluster(clusterUpperName);
-            subscribe(null, event -> {
-                try {
+        if (!LISTENER_SERVICE_MAP.containsKey(clusterUpperName)) {
+            Object lock = CLUSTER_LOCK.computeIfAbsent(clusterUpperName, k -> new Object());
+            if (lock == null) {
```
Comment (a364176773) [inline discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java:139]: 这块多余了,直接下面synchronized即可

File: discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java [hunk 3]
```diff
@@ -109,13 +110,20 @@ public void unregister(InetSocketAddress address) throws Exception {
...
+                .add(listener);
         getEurekaClient(false).registerEventListener(listener);
     }
 
     @Override
     public void unsubscribe(String cluster, EurekaEventListener listener) throws Exception {
-        subscribeListener = false;
+        List<EurekaEventListener> subscribeList = LISTENER_SERVICE_MAP.get(cluster);
+        if (subscribeList != null) {
+            List<EurekaEventListener> newSubscribeList = subscribeList.stream()
```
Comment (a364176773) [inline discovery/seata-discovery-eureka/src/main/java/io/seata/discovery/registry/eureka/EurekaRegistryServiceImpl.java:122]: subscribeList.remove(listener)不就好了吗,为何要过滤,再put会去,map中已经有了该arraylist实例,无需再put一遍

**Review by a364176773 — COMMENTED** [review]
还有个问题就是当前这么改了虽然能用,但是如果在运行中切换不同的集群,其实并没有将先前的集群地址从CLUSTER_ADDRESS_MAP中remove,是否存在内存泄露风险?

**Review by a364176773 — APPROVED** [review]
LGTM

**Review by xingfudeshi — APPROVED** [review]
LGTM

**Pull request merged**